contracts/andromeda_cw721/src/contract.rs

```diff
@@ -0,0 +1,344 @@
+            info.sender.to_string(),
+            Funds::Native(xfer_agreement.amount),
+            to_binary(&ExecuteMsg::TransferNft {
+                token_id: token_id.clone(),
+                recipient: recipient.clone(),
+            })?,
+        )?;
+        let remaining_amount = match remainder {
+            Funds::Native(coin) => coin, //What do we do in the case that the rates returns remaining amount as native funds?
+            Funds::Cw20(..) => panic!("Remaining funds returned as incorrect type"),
+        };
+        resp = resp.add_submessages(msgs).add_events(events);
+        resp = resp.add_submessage(SubMsg::new(CosmosMsg::Bank(BankMsg::Send {
+            to_address: token.owner.clone().into(),
+            amount: vec![remaining_amount],
+        })));
+    }
+
+    check_can_send(deps.as_ref(), env, info, &token)?;
+    token.owner = deps.api.addr_validate(recipient.as_str())?;
+    token.approvals = vec![];
+    token.extension.transfer_agreement = None;
+    token.extension.pricing = None;
+    contract
+        .tokens
+        .save(deps.storage, token_id.as_str(), &token)?;
+
+    Ok(resp)
+}
+
+fn check_can_send(
+    deps: Deps,
+    env: Env,
+    info: MessageInfo,
+    token: &TokenInfo<TokenExtension>,
+) -> Result<(), ContractError> {
+    require(!token.extension.archived, ContractError::TokenIsArchived {})?;
+    // owner can send
+    if token.owner == info.sender {
+        return Ok(());
+    }
+
+    // token purchaser can send
+    if let Some(agreement) = token.extension.transfer_agreement.clone() {
+        require(
+            has_coins(&info.funds, &agreement.amount),
+            ContractError::InsufficientFunds {},
+        )?;
+        if agreement.purchaser == info.sender {
+            return Ok(());
+        }
+    }
+
+    // any non-expired token approval can send
+    if token
+        .approvals
+        .iter()
+        .any(|apr| apr.spender == info.sender && !apr.is_expired(&env.block))
+    {
+        return Ok(());
+    }
+
+    // operator can send
+    let op = AndrCW721Contract::default()
+        .operators
+        .may_load(deps.storage, (&token.owner, &info.sender))?;
+    match op {
+        Some(ex) => {
+            if ex.is_expired(&env.block) {
+                Err(ContractError::Unauthorized {})
+            } else {
+                Ok(())
+            }
+        }
+        None => Err(ContractError::Unauthorized {}),
+    }
+}
+
+fn execute_update_transfer_agreement(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    token_id: String,
+    agreement: Option<TransferAgreement>,
+) -> Result<Response, ContractError> {
+    let contract = AndrCW721Contract::default();
+    let mut token = contract.tokens.load(deps.storage, &token_id)?;
+    require(token.owner == info.sender, ContractError::Unauthorized {})?;
+    require(!token.extension.archived, ContractError::TokenIsArchived {})?;
+    if let Some(xfer_agreement) = agreement.clone() {
```

Comment:
I think this clone is unnecessary as we are borrowing in the body and not taking ownership, should be able to replace with `&agreement` but I could be wrong.